finagle-redis/src/main/scala/com/twitter/finagle/redis/ScriptCommands.scala

```diff
@@ -0,0 +1,147 @@
+package com.twitter.finagle.redis
+
+import com.twitter.finagle.redis.protocol._
+import com.twitter.finagle.redis.util.ReplyFormat
+import com.twitter.util.Future
+import org.jboss.netty.buffer.{ChannelBuffers, ChannelBuffer}
+
+trait Scripts { self: BaseClient =>
+  val handleUnit: PartialFunction[Reply, Future[Unit]] = {
+    case StatusReply(_) | EmptyBulkReply() => Future.Unit
+  }
+
+  val handleLong: PartialFunction[Reply, Future[Long]] = {
+    case IntegerReply(n) => Future.value(n)
+  }
+
+  val handleBool: PartialFunction[Reply, Future[Boolean]] = {
+    case IntegerReply(n)  => Future.value(n == 1)
+    case EmptyBulkReply() => Future.False
+    case StatusReply(_)   => Future.True
+  }
+
+  val handleBuffer: PartialFunction[Reply, Future[ChannelBuffer]] = {
+    case BulkReply(message) => Future.value(message)
+    case EmptyBulkReply() => Future.value(ChannelBuffers.EMPTY_BUFFER)
+  }
+
+  val handleBuffers: PartialFunction[Reply, Future[Seq[ChannelBuffer]]] = {
+    case MBulkReply(messages) => Future.value(ReplyFormat.toChannelBuffers(messages))
+    case EmptyMBulkReply()    => Future.Nil
+  }
+
+  def retryOnNoScript[T](retry: () => T): PartialFunction[Reply, T] = {
+    // NOTE: we cannot make retry call-by-value.
+    // if we make retry call-by-value, the evalXXX() request will happen immediately
+    // when we construct the partial function, thus evalXXX() will be executed before evalShaXXX().
+    // NOTE: so we need call-by-need. but if we use "retry: => T",
+    // when we debug the program, the debugger will force evaluate retry EVERY TIME
+    // when retryOnNoScript is the current frame, resulting in confusion,
+    // especially when retry has side-effects.
+    // so we make retry: () => T an explicity function to prevent the debugger working too eagerly
+    case ErrorReply(message) if message.startsWith("NOSCRIPT") =>
+      retry()
+  }
+
+  // NOTE: because there is no easy way of telling the return type of the passed script,
+  // we rely on the user to tell us the return type.
+  // Thus the different eval* and evalSha* methods correspond to different expected return types.
+  // TODO: currently if the actual reply is different from the told return type,
+  // an IllegalStateException will be thrown. We may want to use a more specific exception for this case.
+  def evalUnit(script: ChannelBuffer, keys: Seq[ChannelBuffer], argv: Seq[ChannelBuffer]): Future[Unit] = {
+    doRequest(Eval(script, keys, argv))(handleUnit)
+  }
+
+  // NOTE: all evalSha* methods take an additional script: Option[ChannelBuffer] parameter,
```

Comment:
rm NOTE
if you expect end-users to want to know this, make this scaladoc.  not everyone reads the code, but most folks read scaladoc.